src/ast/ast_validations.c

```diff
@@ -1638,6 +1628,40 @@ static AST_Validation _AST_Validate_ParseResultRoot(RedisModuleCtx *ctx,
 	return AST_VALID;
 }
 
+static AST_Validation _AST_ValidateUnionQuery(AST *mock_ast, char **reason) {
+	// Verify that the UNION clauses and the columns they join are valid.
+	AST_Validation res = _ValidateUnion_Clauses(mock_ast, reason);
+	if(res != AST_VALID) return res;
+
+	// Each self-contained query delimited by a UNION clause has its own scope.
+	uint *query_scopes = AST_GetClauseIndices(mock_ast, CYPHER_AST_UNION);
+	uint union_count = array_len(query_scopes);
+	AST *scoped_ast;
+	uint scope_end;
+	uint scope_start = 0;
+	for(uint i = 0; i < union_count; i ++) {
+		scope_end = query_scopes[i];
+		// Make a sub-AST containing only the clauses in this scope.
+		scoped_ast = _NewMockASTSegment(mock_ast->root, scope_start, scope_end);
+		res = _ValidateScopes(scoped_ast, reason);
+		AST_Free(scoped_ast);
+		if(res != AST_VALID) goto cleanup;
+
+		// Update the starting index of the scope for the next iteration..
+		scope_start = scope_end;
+	}
+
+	// Build and test the final scope (from the last UNION to the last clause)
+	scope_end = cypher_ast_query_nclauses(mock_ast->root);
+	scoped_ast = _NewMockASTSegment(mock_ast->root, scope_start, scope_end);
```

Comment:
How about appending `cypher_ast_query_nclauses(mock_ast->root)` to `query_scopes` that way the for loop can handle the last clause.

Reply:
Good idea, thanks!